README.md
# Proforma HTML Renderer

[![Gem Version](https://badge.fury.io/rb/proforma-html-renderer.svg)](https://badge.fury.io/rb/proforma-html-renderer) [![Build Status](https://travis-ci.org/bluemarblepayroll/proforma-html-renderer.svg?branch=master)](https://travis-ci.org/bluemarblepayroll/proforma-html-renderer) [![Maintainability](https://api.codeclimate.com/v1/badges/ea2b788b45fcc7d0a183/maintainability)](https://codeclimate.com/github/bluemarblepayroll/proforma-html-renderer/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/ea2b788b45fcc7d0a183/test_coverage)](https://codeclimate.com/github/bluemarblepayroll/proforma-html-renderer/test_coverage) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

[Proforma](https://github.com/bluemarblepayroll/proforma) ships with a basic plain-text renderer.  Luckily we can plug in different rendering engines to provide additional business value.  This library gives Proforma the ability to render HTML instead of simple plain-text.

## Installation

To install through Rubygems:

````
gem install install proforma-html-renderer
````

You can also add this to your Gemfile:

````
bundle add proforma-html-renderer
````

Note: If you are using bundler for auto-requiring then you need to specify as:

```
gem 'proforma-html-renderer', require: 'proforma/html_renderer'
```

## Examples

### Connecting to Proforma Rendering Pipeline

To use this plugin within Proforma:

1. Install [Proforma](https://github.com/bluemarblepayroll/proforma)
2. Install this library
3. Require both libraries
4. Pass in an instance of Proforma::HtmlRenderer into the Proforma#render method

````ruby
require 'proforma'
require 'proforma/html_renderer'

data = [
  { id: 1, name: 'Chicago Bulls' },
  { id: 2, name: 'Indiana Pacers' },
  { id: 3, name: 'Boston Celtics' }
]

template = {
  title: 'nba_team_list',
  children: [
    { type: 'Header', value: 'NBA Teams' },
    { type: 'Separator' },
    {
      type: 'DataTable',
      columns: [
        { header: 'Team ID #', body: '$id' },
        { header: 'Team Name', body: '$name' },
      ]
    }
  ]
}

documents = Proforma.render(data, template, renderer: Proforma::HtmlRenderer.new)
````

The `documents` attribute will now be an array with one object:

```ruby
expected_documents = [
  {
    contents: "...", # HTML Data
    extension: '.html',
    title: 'nba_team_list'
  }
]
```

The `contents` attribute will contain the HTML.

### Customization

All options are passed into Proforma::HtmlRenderer#initialize as a hash:

Name                 | Default
-------------------- | -------
bold_weight          | bold
font_family          | sans-serif
header_font_size     | 18
line_height_increase | 5
text_font_size       | 13

These options will be used during the rendering process.

## Contributing

### Development Environment Configuration

Basic steps to take to get this repository compiling:

1. Install [Ruby](https://www.ruby-lang.org/en/documentation/installation/) (check proforma-html-renderer.gemspec for versions supported)
2. Install bundler (gem install bundler)
3. Clone the repository (git clone git@github.com:bluemarblepayroll/proforma-html-renderer.git)
4. Navigate to the root folder (cd proforma)
5. Install dependencies (bundle)

### Running Tests

To execute the test suite run:

````
bundle exec rspec spec --format documentation
````

Alternatively, you can have Guard watch for changes:

````
bundle exec guard
````

Also, do not forget to run Rubocop:

````
bundle exec rubocop
````

### Publishing

Note: ensure you have proper authorization before trying to publish new versions.

After code changes have successfully gone through the Pull Request review process then the following steps should be followed for publishing new versions:

1. Merge Pull Request into master
2. Update `lib/proforma/html_renderer/version.rb` using [semantic versioning](https://semver.org/)
3. Install dependencies: `bundle`
4. Update `CHANGELOG.md` with release notes
5. Commit & push master to remote and ensure CI builds master successfully
6. Run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## License

This project is MIT Licensed.


CHANGELOG.md
# 1.0.4 (September 9th, 2019)

#### Fixes

* A table should always be rendered, even if it has no rows.  This puts this library's table rendering on par with proforma-pdf-renderer, where it renders empty tables as well.

# 1.0.3 (September 9th, 2019)

#### Fixes

* Fixed bug rendering table (empty message is not a Table notion, it is a DataTable notion.)

# 1.0.2 (June 14th, 2019)

#### Additions

* Added surrounding <div> element encapsulating all child elements.  This should help with embedding the contents inside consumer web applications while still using the provided custom styles while preventing cross-contamination.

# 1.0.1 (June 6th, 2019)

#### Additions

* Added additional right padding to middle pane value cells.

#### Fixes

* Fixed bug with Banner objects not rendering.

# 1.0.0 (June 6th, 2019)

Initial release.


spec/spec_helper.rb
# frozen_string_literal: true

#
# Copyright (c) 2018-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require 'fileutils'
require 'pry'
require 'pry-byebug'
require 'yaml'

unless ENV['DISABLE_SIMPLECOV'] == 'true'
  require 'simplecov'
  require 'simplecov-console'

  SimpleCov.formatter = SimpleCov::Formatter::Console
  SimpleCov.start do
    add_filter %r{\A/spec/}
  end
end

require './lib/proforma/html_renderer'


spec/rendering_helper.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require './lib/proforma/html_renderer'

# This class can render based on a conventional directory structure of input files.
class RenderingHelper
  class << self
    def folders(path)
      Dir.entries(path).select do |entry|
        File.directory?(File.join(path, entry)) && !%w[. ..].include?(entry)
      end
    end
  end

  attr_reader :path

  def initialize(path)
    @path = path
  end

  def render
    images    = images_hash(images_dir)
    config    = yaml_read(config_filename)
    data      = data_merge(config['data'], images)
    template  = config['template']  || {}
    evaluator = config['evaluator'] || Proforma::HashEvaluator.new
    renderer  = config['renderer']  || Proforma::HtmlRenderer.new

    Proforma.render(
      data,
      template,
      evaluator: evaluator,
      renderer: renderer
    )
  end

  def render_and_save
    output_dir = File.join(path, 'output')

    FileUtils.mkdir_p(output_dir)

    render.each_with_index do |output, index|
      name = output_filename(output, index)
      full_path = File.join(output_dir, name)

      IO.write(full_path, output.contents)
    end
  end

  def output_filename(output, index)
    name_without_extension = [
      index.to_s,
      output.title.to_s
    ].reject(&:empty?).join('.')

    "#{name_without_extension}#{output.extension}"
  end

  private

  def images_dir
    File.join(path, 'images')
  end

  def config_filename
    File.join(path, 'config.yml')
  end

  def yaml_read(file)
    YAML.safe_load(File.open(file))
  end

  def data_merge(data, hash)
    if data.is_a?(Array)
      data.map do |record|
        record.merge(hash)
      end
    elsif data.is_a?(Hash)
      data.merge(hash)
    end
  end

  def images_hash(path)
    images_glob = File.join(path, '*')

    Dir[images_glob].map do |image_filename|
      key = File.basename(image_filename, File.extname(image_filename))
      [key, File.open(image_filename, 'rb')]
    end.to_h
  end
end


spec/proforma/html_renderer_spec.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require 'rendering_helper'
require 'spec_helper'

describe Proforma::HtmlRenderer do
  specify 'snapshot has been un-modified' do
    snapshot_dir = File.expand_path(File.join(__dir__, '..', 'fixtures', 'snapshot'))
    rendering_helper = RenderingHelper.new(snapshot_dir)

    actual_output = rendering_helper.render.map.with_index do |output, index|
      [
        rendering_helper.output_filename(output, index),
        output.contents
      ]
    end.to_h

    output_glob = File.join(snapshot_dir, 'output', '*')

    expected_output = Dir[output_glob].map do |filename|
      [File.basename(filename), File.open(filename, 'rb').read]
    end.to_h

    expect(actual_output).to eq(expected_output)
  end
end


lib/proforma/html_renderer/writer.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

module Proforma
  class HtmlRenderer
    # This class can render ERB templates.
    class Writer
      # This nested class is an internal intermediate only used by the encapsulating Writer class.
      # It is simply used to convert an OpenStruct object to a Binding object which can be used
      # for ERB template rendering.
      class ErbBinding < OpenStruct
        def expose_binding
          binding
        end
      end

      BACK_DIR      = '..'
      ERB_OPTIONS   = '-'
      TEMPLATE_DIR  = 'templates'

      def initialize(directory: nil, files: {})
        @directory  = directory || default_directory
        @files      = to_erb_hash(files)
      end

      def render(name, context = {})
        erb_template(name).result(hash_to_binding(context))
      end

      private

      attr_reader :directory, :files

      def hash_to_binding(hash)
        ErbBinding.new(hash).expose_binding
      end

      def erb_template(name)
        files[name.to_s] ||= read(name)
      end

      def read(name)
        text = IO.read(path(name)).chop
        ERB.new(text, nil, ERB_OPTIONS)
      end

      def default_directory
        File.join(
          __dir__,
          BACK_DIR,
          BACK_DIR,
          BACK_DIR,
          TEMPLATE_DIR
        )
      end

      def path(name)
        File.join(directory, name.to_s)
      end

      def to_erb_hash(hash)
        (hash || {}).map do |(name, file)|
          [
            name.to_s,
            file.is_a?(ERB) ? file : ERB.new(file)
          ]
        end.to_h
      end
    end
  end
end


lib/proforma/html_renderer/object_writer.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require_relative 'writer'

module Proforma
  class HtmlRenderer
    # Each object should have its own ERB template.  This class understands the connection of
    # objects and templates and knows how to render the template for the respective object
    # passed in.
    class ObjectWriter
      attr_reader :names_by_class, :writer

      def initialize(names_by_class: {}, writer: Writer.new)
        @names_by_class = names_by_class
        @writer         = writer
      end

      def render(object, options = {})
        writer.render(
          resolve_name(object),
          object: object,
          options: options
        )
      end

      private

      def resolve_name(object)
        names_by_class[object.class].tap do |name|
          raise ArgumentError, "object cannot be rendered: #{object.class}" unless name
        end
      end
    end
  end
end


lib/proforma/html_renderer/proforma_writer.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require_relative 'object_writer'

module Proforma
  class HtmlRenderer
    # This class connects each Proforma object to its proper template for rendering.
    class ProformaWriter
      extend Forwardable

      # The singleton is provided as a global helper since we want to expose the same
      # static functionality across the entire app.  The real value is the caching of the
      # Writer's files so we are only reading each template the initial time.
      class << self
        extend Forwardable

        def_delegators :instance, :render

        def instance
          @instance ||= new
        end
      end

      NAMES_BY_CLASS = {
        Modeling::Banner => 'banner.html.erb',
        Modeling::Header => 'header.html.erb',
        Modeling::Pane => 'pane.html.erb',
        Modeling::Separator => 'separator.html.erb',
        Modeling::Spacer => 'spacer.html.erb',
        Modeling::Table => 'table.html.erb',
        Modeling::Text => 'text.html.erb',
        Prototype => 'prototype.html.erb'
      }.freeze

      attr_reader :object_writer

      def_delegators :object_writer, :render

      def initialize(object_writer: ObjectWriter.new(names_by_class: NAMES_BY_CLASS))
        @object_writer = object_writer
      end
    end
  end
end


lib/proforma/html_renderer.rb
# frozen_string_literal: true

#
# Copyright (c) 2019-present, Blue Marble Payroll, LLC
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

require 'base64'
require 'erb'
require 'forwardable'
require 'ostruct'
require 'proforma'

require_relative 'html_renderer/proforma_writer'

module Proforma
  # This main class to use as a Proforma renderer.
  class HtmlRenderer
    EXTENSION = '.html'

    DEFAULT_OPTIONS = {
      bold_weight: 'bold',
      font_family: 'sans-serif',
      header_font_size: 18,
      line_height_increase: 5,
      text_font_size: 13
    }.freeze

    attr_reader :options

    def initialize(options = {})
      @options = OpenStruct.new(DEFAULT_OPTIONS.merge(options))
    end

    def render(prototype)
      contents = ProformaWriter.render(prototype, options)

      Proforma::Document.new(
        contents: contents,
        extension: EXTENSION,
        title: prototype.title
      )
    end
  end
end
